docs/src/docs/asciidoc/developer/clustered.adoc

```diff
@@ -0,0 +1,174 @@
+--------------------------------------------------------------------------------
+
+The structure above is interpreted by the client as meaning the current value of
+the k1 mapping is `g(f({k1, v1}))`, and that of the k2 mappings is
+`f({k2, v1}`.  More generally the current value of a mapping is always equal
+to a recursive application of the functions in its sequence (treating the
+'value' as an initial generator function).
+
+At this point we define a critical client side operation:
+
+[source]
+----
+value resolve(chain, key) {
+  resolved = [];
+  transform = () -> {};
+  for (function f : chain)) {
+    if (f.for(key)) {
+      transform = transform.andThen(f);
+    } else {
+      resolved.add(f);
+    }
+  }
+  value = transform.apply();
+  resolved.add(value)
+
+  asynchronously(server.replace(key.hash, chain, resolved));
+
+  return value;
+}
+----
+
+Invocation of this operation resolves the current value of the key, and
+asynchronously updates the servers representation:
+
+["ditaa", "resolve-operation", "png"]
+--------------------------------------------------------------------------------
+      +----+    +----------------------------------------------------------+
+      |    |    | +------+ /----------\ +------+ /----------\ /----------\ |
+      |hash| => | |k1, v1| |k1, f(k,v)| |k2, v1| |k1, g(k,v)| |k2, f(k,v)| |
+      |    |    | +------+ \----------/ +------+ \----------/ \----------/ |
+      +----+    +----------------------------------------------------------+
+                                             |
+
+                                     resolve(chain, k1)
+
+                                             |
+                                             V
+                            +--------------------------------+
+                            | +------+ /----------\ +------+ |
+                            | |k2, v1| |k2, f(k,v)| |k1, v2| |
+                            | +------+ \----------/ +------+ |
+                            +--------------------------------+
+                                             |
+
+                                     resolve(chain, k2)
+
+                                             |
+                                             V
+                                   +-------------------+
+                                   | +------+ +------+ |
+                                   | |k1, v2| |k2, v2| |
+                                   | +------+ +------+ |
+                                   +-------------------+
+--------------------------------------------------------------------------------
+
+== Example Operations
+
+[source]
+----
+void put(key, value) {
+  server.append(key.hash, function(put(key, value)));
+}
+
+value get(key) {
+  return resolve(server.get(key.hash), key);
+}
+
+value putIfAbsent(key, value) {
+  return resolve(server.getAndAppend(key.hash, operation(putIfAbsent(key, value))), key);
+}
+----
```

Comment:
As a reminder to myself: depending on how costly it would be to calculate the list size on the server, it could be the return value of the `append()` method then it would be up to the client to have the logic to proactively call `resolve()` or not. Otherwise, the server could hint the client by returning a boolean flag instead.